File: src/utils/api/counts.ts
import API_ENDPOINTS from "src/config/api";

export const getCounts = async (
  access_token: string,
  data_stream_id: string,
  data_stream_route: string,
  date_start?: string,
  date_end?: string,
  ext_event?: string
): Promise<Response> => {
  const params = new URLSearchParams();

  if (data_stream_id) params.append("data_stream_id", data_stream_id);
  if (data_stream_route) params.append("data_stream_route", data_stream_route);
  if (date_start) params.append("date_start", date_start);
  if (date_end) params.append("date_end", date_end);
  if (ext_event) params.append("ext_event", ext_event);

  const url = `${API_ENDPOINTS.counts}?${params.toString()}`;

  const response = await fetch(url, {
    method: "GET",
    headers: {
      "Content-Type": "application/json",
      Authorization: "Bearer " + access_token,
    },
  }).catch();

  return response;
};

File: api/src/main/kotlin/dexportal/routes/psApi/Counts.kt
package dexportal.routes.psApi

import com.apollographql.apollo3.ApolloClient
import com.apollographql.apollo3.api.toJson
import com.dexportal.CountsQuery
import dexportal.utils.InvalidDateFormatException
import dexportal.utils.processDateString

import io.ktor.http.*
import io.ktor.server.application.*
import io.ktor.server.response.*
import io.ktor.server.routing.*

fun Route.countsRoute(apolloClient: ApolloClient) {
    route("/report-counts") {
        get {
            val dataStream = call.request.queryParameters["data_stream_id"]
            if (dataStream == null || dataStream == "") {
                call.respond(HttpStatusCode.BadRequest, "Datastream name must be provided")
                return@get
            }

            val dataRoute = call.request.queryParameters["data_stream_route"] ?: ""

            val dateStart = call.request.queryParameters["date_start"]
            val dateEnd = call.request.queryParameters["date_end"]
            if (dateStart == null || dateStart == "" || dateEnd == null || dateEnd == "") {
                call.respond(HttpStatusCode.BadRequest, "Date start and date end must be provided")
                return@get
            }

            val authToken = call.request.headers["Authorization"]
            if (authToken == null || authToken == "") {
                call.respond(HttpStatusCode.BadRequest, "Authorization token must be provided")
                return@get
            }

            val cleanedDateStart: String?
            val cleanedDateEnd: String?

            try {
                cleanedDateStart = processDateString(dateStart)
                cleanedDateEnd = processDateString(dateEnd)
            } catch (e: InvalidDateFormatException) {
                call.respond(HttpStatusCode.BadRequest, "${e.message}")
                return@get
            }

            try {
                val response = apolloClient.query(
                    CountsQuery(
                        dataStream, dataRoute, cleanedDateStart, cleanedDateEnd
                    )
                ).addHttpHeader("Authorization", authToken).execute()
                val jsonData = response.data?.toJson()
                call.respond(HttpStatusCode.OK, jsonData.toString())
            } catch (e: Exception) {
                call.respond(
                    HttpStatusCode.InternalServerError,
                    "An error occurred making the request to PS-API: ${e.message}"
                )
            }
        }
    }
}
